trade.lic

```diff
@@ -0,0 +1,258 @@
+        contract.destination_clerk = @town_data[contract.destination_town]['shipment_clerk']['id']
+        contract.destination_minister = @town_data[contract.destination_town]['trade_minister']['id']
+      elsif line =~ payment_re
+        # This is the last line of the contract so break when finished
+        contract.payment, contract.currency = line.match(payment_re).captures
+        break
+      elsif line =~ origin_re
+        contract.origin_town = line.match(origin_re).captures.first
+        contract.origin_clerk = @town_data[contract.origin_town]['shipment_clerk']['id']
+        contract.origin_minister = @town_data[contract.origin_town]['trade_minister']['id']
+        contract.origin_outpost = @town_data[contract.origin_town]['trader_outpost']['id']
+      elsif line =~ /This contract has yet to be/
+        contract.presented = false
+      elsif line =~ /The contract is now useless since it has expired\./
+        contract.expired = true
+        break
+      end
+    end
+    if $debug
+      echo "Contract origin: #{contract.origin_town}"
+      echo "Contract origin clerk room: #{contract.origin_clerk}"
+      echo "Contract origin minister room: #{contract.origin_minister}"
+      echo "Contract destination town: #{contract.destination_town}"
+      echo "Contracted presented to clerk? #{contract.presented}"
+    end
+    return contract
+  end
+
+  def recall_caravan
+    recall_messages = [
+        /You seem to recall that you left your (?<description>.*) (?<name>.*) right behind you/,
+        /You recall that your (?<description>.*) (?<name>.*) should be located in/,
+        /You don't recall where you left your caravan\./,
+        /You don't recall having a caravan\./
+    ]
+    response = bput("recall caravan", recall_messages)
+    echo("Response: #{response}") if $debug
+    case response
+    when /You don't recall where you left your caravan/, /You don't recall having a caravan/
+      echo("No caravan found") if $debug
+    when /You seem to recall that you left your (?<description>.*) (?<name>.*) right behind you/
+      /You seem to recall that you left your (?<description>.*) (?<name>.*) right behind you/ =~ response
+      @caravan_noun = name
+      echo "@caravan_noun: #{@caravan_noun}" if $debug
+      @caravan_adjective = description
+      echo "@caravan_description: #{@caravan_adjective}" if $debug
+    when /You recall that your (?<description>.*) (?<name>.*) should be located in/
+      /You recall that your (?<description>.*) (?<name>.*) should be located in/ =~ response
+      @caravan_noun = name
+      echo "@caravan_noun: #{@caravan_noun}" if $debug
+      @caravan_adjective = description
+      echo "@caravan_description: #{@caravan_adjective}" if $debug
+    end
+  end
+
+  def compute_path(destination)
+    previous, _shortest_paths = Map.dijkstra(Room.current.id, destination)
+    path = [destination]
+    path.push(previous[path[-1]]) until previous[path[-1]].nil?
+    path.reverse!
```

Comment:
Maybe I'm reading this wrong, but does this build the path in reverse order, then reverse it? Can you reference `path[0]` and build it in-order?

Reply:
I cant think of a good way to build it starting from the source, at least not by looking up the index of the room pointing to it. There might be several indices that point to the starting room for example

```
[test: source: 1005]
[test: destination: 895]
[test: previous[1006]: 1005]
[test: previous.index(1005): 1004]
[test: previous.index(source): 1004]
[test: forward_path: [1005, 1004, 1003, 1001, 1000, 999, 998, 997, 996, 995, 992]]
[test: backward_path: [1005, 1006, 1007, 1642, 1626, 1625, 1623, 1619, 1620, 1621, 806, 805, 807, 808, 809, 810, 770, 771, 772, 862, 905, 906, 904, 907, 903, 902, 895]]
```

Reply:
The last pseudo code entry discusses the same method of reconstructing the path by reverse iteration as well: 

https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm#Pseudocode

Reply:
👍